### ISSUE TITLE
Jupyter over remote ssh sometimes becomes slow and/or unresponsive

### ISSUE THREAD

**Issue by murphycj2** [microsoft/vscode#138784]
## Environment data

-   VS Code version: 1.62.3 (Universal)
-   Jupyter Extension version (available under the Extensions sidebar): v2021.10.1101450599
-   Python Extension version (available under the Extensions sidebar): v2021.11.1422169775
-   OS (Windows | Mac | Linux distro) and version: Mac
-   Python and/or Anaconda version: 3.7.12
-   Type of virtual environment used (N/A | venv | virtualenv | conda | ...): conda
-   Jupyter server running: Remote

## Expected behaviour

I expect code cells to begin executing immediately and (for simple code snippets) to finish executing immediately (e.g. print('hello'))

## Actual behaviour

Once the jupyter notebook starts to contain a non-trivial amount of stuff, I start to observe the following behaviors: (1) code cells that contain simple tasks (e.g. `print('hello')`) start taking several seconds or more to complete, (2) it may take several seconds or even minutes for vscode to even visually show that it will start executing a code cell, and (3) in severe cases, the whole vscode editor will become unresponsive and I will need to force shut it down.

## Steps to reproduce:

Below is code for a jupyter notebook that contains a minimal example. The problems start to appear once you've cycled through the code cells 2 or 3 times. I think one of the main contributing factors to this issue are just notebooks that have a lot of "stuff" in them. To demonstrate that, the last code block prints 50 scatter plots as pngs, which the native jupyter lab server over the browser handles just fine... but vscode seems to have issues. Moreover, in this simple example I have observed instances where it starts to take a while to save the notebook and the editor may even become unresponsive. 

Other things to note:
- I am running this jupyter notebook on an AWS EC2 instance with ~90GB memory 48 CPUs (using a AWS linux ami: `ami-083ac7c7ecf9bb9b0`).
- I have not yet observed these same issues when I have the kernel running on my local mac machine.

<details>

<summary>Jupyter example
</summary>

<p>

```
# %%
# %load_ext autoreload
# %autoreload 2

import pandas as pd
import plotly.express as px
import numpy as np
import plotly.io as pio
pio.renderers.default='jupyterlab+notebook'
pd.set_option("display.max_columns", 20)

# %%
data = pd.DataFrame(np.random.random(size=(1000,1000)))

# %%
data

# %%
print('hello')

# %%
for i in range(50):
    fig = px.scatter(x=data[i], y=data[i+1])
    fig.show(renderer='png')
    
```

</p>
</details>

## Logs

Here is a screenshot where it takes almost 3 seconds for a simple print statement. I've seen worse in some of my code related to real projects, though. This is just what I was able to reproduce with a minimal example.

<img width="1593" alt="Screen Shot 2021-11-22 at 4 48 27 PM" src="https://user-images.githubusercontent.com/52719928/142940256-73fbc9d9-6b80-413b-bdc4-63adfa79bfc0.png">

**Comment by NixGD** [comment]
Also having this issue with jupyter over ssh, in a notebook that has a decent number of plots.

**Comment by MohamedAliRashad** [comment]
This issue happened with me also (especially when i am insider a docker container inside the remote machine)

**Comment by StephaneKazmierczak** [comment]
Same problem, sometimes it works fine, sometimes it is unusable

**Comment by greazer** [comment]
Possibly add logging for every message that's only turned on if the user explicitly is told to (for debugging purposes).

**Comment by erdaltoprak** [comment]
Same on my side, any way to hints to resolve this ?

`Roundtrip latency: 4.250ms
Up: 824.7 Mbps
Down: 490.5 Mbps


Authority: ssh-remote+jupyter
Roundtrip latency: 2071.960ms
Up: 17.4 Mbps
Down: 188.4 Mbps
`

**Comment by svechinsky** [comment]
experiencing similar issues when working on jupyter via ssh to an ubuntu server from an ubuntu pc

**Comment by Jumabek** [comment]
ALso had the issue when using remote SSH
- aws
- azure

**Comment by wyf0912** [comment]
Still the same problem

**Comment by johnros-pagaya** [comment]
I also experience the same. I observed that:
- It happens with Jupyter notebooks, but not in the Interactive window. 
- It does not happen if you connect to a Jupyter kernel that is not managed by VScode. 
- It happens when your notebooks has many figures. 

I hope these may help diagnose the problem.

**Comment by vdplasthijs** [comment]
Same here..

**Comment by Jumabek** [comment]
As an additional info to what @johnros-pagaya enlightened us.

Jupyter execution on remote ssh machine slows down (maybe gets stuck is the right description) when I notebook size is large (64 MB).

I tested smaller sized 1MB notebook and it executed just fine

**Comment by DonJayamanne** [comment]
> when notebook size is large (64 MB).

Would it be possible to share such a notebook? Is the size 64 MB with or without ouput in the notebook?
If its a large file you could optionally email it to one of us (us = Microsoft maintainers of VS Code)

**Comment by Jumabek** [comment]
@DonJayamanne sorry that is confidential work for the time being.

**Comment by movy** [comment]
Not sure if it's the same issue, but even pasting python code in a small simple cell  lags lots of times, i.e. I can press cmd+v 10 times and nothing would happen then suddely it pastes 2-3 copies of the same code.. Latest VSCode @ remote Ubuntu w/ Juputer Lab, accessed from local OSX.

**Comment by CaoHoangTung** [comment]
Experiencing the same issue as well. Networking is not a problem since I connect to the notebook's machine on a local network

**Comment by emilyhunt** [comment]
I've also been having this issue lately although I think it only happens in certain circumstances. I can't share the notebook I'm working with at the moment as it's private (sorry!), but it does have many matplotlib figure outputs and dataframe outputs. The notebook itself is only 3.6 MB in size. I'm running the notebook on a Linux server with Python 3.9.7, specifically I'm connecting to the machine remotely with "Connect to Host" to edit and run code on the server remotely. Latest version of VS Code and all extensions.

It seems connecting on the same network as the machine with my Linux laptop doesn't result in any issues, and I haven't noticed any problems with running the notebook from there. However, connecting from a different network when working from home causes the same notebook to be very laggy, including lagging for 30+ seconds on simple code cells containing e.g. just a print statement. I don't see how network speed/my connection to the server from home could be the issue (12ms ping, ~5 Mbps upload / ~50 Mbps download connections). Maybe there's something in the different OSes having different results, could also be unrelated though. I should try connecting with a Linux machine not on the same network and see if I can reproduce the results.

**Comment by Gorluxor** [comment]
> I also experience the same. I observed that:
> 
> * It happens with Jupyter notebooks, but not in the Interactive window.
> * It does not happen if you connect to a Jupyter kernel that is not managed by VScode.
> * It happens when your notebooks has many figures.
> 
> I hope these may help diagnose the problem.

Same problem here, just a correction it is not dependent on the figure number or commands difficulty, `print("Hello")` will take to execute 30 to 40s, and yes with a separate instance of Jupyter Lab server (not managed by VScode). Directly connecting to Jupyterlab over ssh port forwarding order of magnitude faster. 

Workflow: Windows to Linux remote SSH (cluster) > Another instance in the cluster used to set up JupyterLab instance > Change Jupyter server to remote of the previously created instance > execution super slow (even delay between starting and finishing) ~ 30 second delay followed up by ~ 40 s execution
Connecting directly from Windows > tunnel to JupyterLab > ~ few seconds execution
Note: Both the Jupyter Lab instance and the initial remote ssh share access to the file system, just different hardware. 
Also after changing to remote, some shenanigans happen with the vscode, unable to open files, etc, had to restart everything to make it work again.

**Comment by ckrapu** [comment]
I've had the same issue with VScode and Jupyter - notebooks becoming unresponsive after a few cells are used. The console shows continuous (every 20 seconds) reconnecting. [Here's the console output ](https://gist.github.com/ckrapu/df222a618d4f926b135f68d9bc1695aa)with the level set to "trace". I can share a relevant `.ipynb` file if it is helpful.

**Comment by X-funbean** [comment]
I've had the same slow issue with VScode and Jupyter when connecting to a remote docker container via ssh, and I've also observed that ssh.exe keeps producing tremendous streams

**Comment by roblourens** [comment]
If someone can share a notebook that triggers this issue (lots of outputs, graphs with lots of data, etc?) that would be helpful, thanks

**Comment by Tai-Zhou** [comment]
I noticed that if I switch the format from svg to png (by changing the setting `jupyter.generateSVGPlots`), the notebook would run smoother.

---
My network:

Roundtrip latency: 5.520ms
Up: 2898.6 Mbps
Down: 3007.5 Mbps


Authority: ssh-remote+******
Roundtrip latency: 21.320ms
Up: 25.8 Mbps
Down: 385.0 Mbps

---
A test notebook (1KB zip file), need `numpy`, `matplotlib` and `pandas` to run:

[Issue 138784.ipynb.zip](https://github.com/microsoft/vscode/files/10038299/Issue.138784.ipynb.zip)

**Comment by shnitzel6** [comment]
Having the same exact issue and it renders the jupyter unresponsive:
![image](https://user-images.githubusercontent.com/97987330/203349093-e2fe1582-90be-4f1b-b40c-6675d6f308a3.png)
And it's still running.
A temporary solution I found is to click on "Clear outputs for all cells" and more often than it works smooth again. Doesn't work if the cell doesn't stop executing though...

**Comment by TTTPOB** [comment]
I think notebook size may be the reason. https://github.com/microsoft/vscode-jupyter/issues/12284

**Comment by shnitzel6** [comment]
I resolved it by using a jupyter server running on the remote machine which was not started by VSC. I'll use that until this bug is fixed unfortunately

**Comment by BounharAbdelaziz** [comment]
@lironle6 life saviour! "Clear outputs for all cells" worked smooth as butter (:

**Comment by lironle6** [comment]
@BounharAbdelaziz Happy to help. I saw in one of the other tickets this bug has a synergy with notebook size, and since clearing all outputs reduces notebook size, this might be a temporary fix until the size increases eventually and we need to do it again.
I would love to see this fixed as this bug is the biggest hindrance to me day-to-day vscode experience.

**Comment by vincenzoml** [comment]
> I resolved my issue by closing the JUPYTER: VARIABLES in the terminal below Maybe in my case was an issue with the output variables inside the jupyter section ..

Same here, this issue was driving me crazy. That's the solution indeed, I had a lot of local memory usage in my code, not sure about what was the actual issue.

**Comment by roblourens** [comment]
I pushed a change which should mostly fix this slowness with large outputs. I think there are more optimizations we could do. The issue comes when a save happens, and this blocks the remote extension host process for some period of time. Then a symptom of that might be that an execution appears to hang or take a long time.

Would appreciate anyone trying this out in tomorrow's vscode insiders build.

**Comment by paul-brenner** [comment]
I have been experiencing these slowdowns (specifically on an AWS server) and I don't use the variables view. I definitely found that clearing cell outputs eliminates the hangs. 

I'm confused how this is caused by saving though, because I just found that the notebook I was working on wasn't being saved. I was doing some work tethered to my phone as I rode a bus into remote mountains. Certainly noticed that the bad connectivity meant even more waiting for things to execute than usual. 

Unfortunately, when I got back to stronger internet vscode was complaining it couldn't connect and was hanging when it tried to save. I figured it was fine since I saw @roblourens 's comment and naively interpreted that as meaning I should feel comfortable that my notebook was getting saved each time it hung. Nope. When I finally had to reload window to get things back to working I found that all the code I wrote on the bus was no longer in the notebook (no worries, I had copied it out just in case, so no loss). 

I'm trying the 2023-01-13T07:49:05.588Z insiders build right now (is the "tomorrow's" build that should have the changes?). I think it is still too soon for me to be able to offer an opinion on how much of an improvement this fix might be. It is hard because the problem is erratic: sometimes cells execute immediately and sometimes it takes 20 seconds or longer.

**Comment by roblourens** [comment]
Yeah, that version has the latest fix, you might have hit a different issue. And the issue might still have been kicked off by the notebook _attempting_ to save. I'll leave this issue open to track

>  I think there are more optimizations we could do.

namely, we still send notebook data between the processes more often than we technically have to, and I think that can be improved.

> When I finally had to reload window to get things back to working I found that all the code I wrote on the bus was no longer in the notebook

Sorry about that, glad you had it saved, the intent is that the unsaved changes would have been persisted on the frontend.

**Comment by paul-brenner** [comment]
I can confirm that this change is a big improvement for my use case. If those additional optimizations ever get prioritized that would be amazing and appreciated, but I'm already grateful for this improvement. Thank you!

**Comment by roblourens** [comment]
On second thought, closing this issue, and moving other work to https://github.com/microsoft/vscode/issues/172345

**Comment by vincenzoml** [comment]
I have tried the insiders version on a podman container accessed on a machine over ssh (so I do vscode remote via ssh to the machine and then I use remote containers to access the container).

It does not open a terminal and seems to be spinning forever with the following error in the logs:

2023-01-27 10:40:00.790 [info] [LocalProcess0][resolveAuthority(attached-container,27)][1013ms] waiting...
2023-01-27 10:40:00.818 [error] Shell server failed: Error: stream ended with:4 but wanted:9
	at u (c:\Users\vince\.vscode-insiders\extensions\ms-vscode-remote.remote-containers-0.273.0\dist\extension\extension.js:11:65948)
	at c:\Users\vince\.vscode-insiders\extensions\ms-vscode-remote.remote-containers-0.273.0\dist\extension\extension.js:11:66129
	at s (c:\Users\vince\.vscode-insiders\extensions\ms-vscode-remote.remote-containers-0.273.0\dist\extension\extension.js:14:5342)
	at Socket.<anonymous> (c:\Users\vince\.vscode-insiders\extensions\ms-vscode-remote.remote-containers-0.273.0\dist\extension\extension.js:14:5512)
	at Socket.emit (node:events:538:35)
	at endReadableNT (node:internal/streams/readable:1345:12)
	at process.processTicksAndRejections (node:internal/process/task_queues:83:21)

May this be related to the recent changes or shall I seek for another bug report and eventually open a new one?

**Comment by roblourens** [comment]
Not related to this change- you can open an issue against the Remote Containers extension, I'm not sure how to interpet that  error

**Comment by bo44arov** [comment]
Hello,
I have installed latest Insider's VS Code, maybe it's a bit better now, but in general issue is still there.
Very slooow Jupyter Notebooks on large content.

I'm using jupyter/r-notebook docker image to run on AWS EC2 instance and have 2 options to acces it:
- from browser on 8888 port to web version of Jupyter notebooks. It works lightning fast and responsive.
- over SSH to EC2 instance and then "Attach to running container"

So if I compare those 2 options there is a huge difference in UI when notebook has some data and charts.
For me it feels like VS Code continuously sends data to server. Even dropdown for intellisence takes few seconds to load.

It's so sad that I'm unable to use VS Code Jupyter Notebooks for my work.

But if I create a brand new empty notebook in VS Code - everything works like a charm!

**Comment by rebornix** [comment]
@bo44arov  @paul-brenner  We have added experimental saving logic for Remote SSH, it would be great if you can give this a try and see if it improves the performance on large notebooks

* Install latest VS Code Insiders
* Remote SSH into your remote machine
* Add `"notebook.experimental.remoteSave": true` to your Remote/User settings 
* Reload the window
* Then test the scenarios which used to slow or block the network

Thanks in advance!

**Comment by ma-ji** [comment]
@rebornix The improvements are noticeable, thank you for your great work! However, the latency is still there and appears to be significant from time to time. 

I love this ssh-notebook function, and I believe there must be more folks like me. Hope your team can prioritize this issue. 

Appreciate!

**Comment by bo44arov** [comment]
Hi,
I just checked VS Code Insiders 1.81.0 version with "Remote Save" feature enabled in settings.

I have generated in a loop 30 charts in first cell and after this I run code in a new cell. It still hangs for a while after generating charts for the first run. If I run second cell code again, it executes immediately as expected.

So we still have an issue when a lot of data (or visual data) was generated recently.

I checked the same code in web UI of Jupyter - and it works fast.

Anton.
